drivers: video: Add pwdn pin support on ov2640

yunsmall opened the pull request:
The current ov2640 driver does not support pwdn pin support, so if there is a separate pin to control pwdn, there will be problems. It is necessary to add `zephyr,deferred-init` tags in the device tree to manually initialize, which is inconvenient. Tested on dnesp32s3 from alientek, it works. I also formatted the code with clangd.

avolmat-st reviewed (changes requested):
Hello @yunsmall and thank you for this modification.

There is too many modifications at once in a single commit in order to be able to review it. I understand that there is not only the addition of the pwdn pin but also you've probably ran a tool to automatically fix codying style of the driver.

Looking the modifications in this commit, it seems to be necessary at some place, but in such case it should be done in a dedicated commit with the codying style fixes alone are done since doing several things at once in a commit doesn't allow us to review the real modifications.

Could you update your PR by splitting this into several commits, for ex:
1st commit: addition of the pwdn pin support (hence I think modification in ov2640.c and yaml file)
2nd commit: fix of major codying styles if necessary

The shield / dts for using this sensor on the dnesp32s3 is not part of Zephyr ? If so, I think it should also be updated to add the pwdn-pin now that this is available and allow the sensor to be used.

Thanks a lot for your help.

yunsmall committed:
drivers: video: ov2640: Add pwdn pin support on ov2640.

Add handling for the power-down (pwdn) pin of the OV2640 sensor.

Signed-off-by: Yunjie Ye <yun_small@163.com>

yunsmall commented:
@avolmat-st Thank you for your response for my PR. I did what you suggested to me, which is splitting the commit into two separated commits.
The board dnesp32s3 is an esp32s3 board released by Alientek, which is not a part of Zephyr. And I am also not a member of Alientek, so I suppose that the dts of this board should be added by their team.
Much thanks for this wonderful project.

avolmat-st reviewed (changes requested):
Thanks for the update and sorry for the delay of the review.

Small comments.

Could you also rework the commit messages of your 2 commits. You can have a look at Zephyr Contribution page for information about how to write commit messages:

https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

Thanks a lot.

yunsmall commented on drivers/video/ov2640.c:
```diff
@@ -1008,9 +1016,17 @@ static int ov2640_init(const struct device *dev)
-	const struct ov2640_config *cfg = dev->config;
+#if DT_INST_NODE_HAS_PROP(0, pwdn_gpios)
+	ret = gpio_pin_configure_dt(&cfg->pwdn_gpio, GPIO_OUTPUT_ACTIVE);
+	if (ret) {
```
Applied.

josuah commented on drivers/video/ov2640.c:
```diff
@@ -463,16 +459,14 @@ struct ov2640_data {
+	 .height_min = (height),                                                                   \
+	 .height_max = (height),                                                                   \
+	 .width_step = 0,                                                                          \
+	 .height_step = 0}
```
Here, `clang-format` did not wrap the braces in the conventional style.
This can be controlled by adding a trailing `,` after the `0` on the last line, before the `}`, and re-applying clang-format again.

josuah commented on drivers/video/ov2640.c:
```diff
@@ -1077,8 +1095,14 @@ static int ov2640_init_0(const struct device *dev)
 
-	uint32_t i2c_cfg = I2C_MODE_CONTROLLER |
-					I2C_SPEED_SET(I2C_SPEED_STANDARD);
+#if DT_INST_NODE_HAS_PROP(0, pwdn_gpios)
```
The problem with this is that it only checks for the instance `0` (the first one).
If there are 2 sensors and only the 2nd has `pwdn_gpios`, then this would not work.

You might need something like this:

```suggestion
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
	if (cfg->pwdn_gpios.port != NULL) {
```

The `#if` helps making `cfg->pwdn_gpios` compile.

The `if ()` helps detecting the per-instance presence/absence of a GPIO port.

You might need to replicate this for each occurence of `DT_INST_NODE_HAS_PROP(0, pwdn_gpios)`.

Would that be ok?
Thank you!

josuah reviewed (commented):
About the commit message: would it be possible to wrap the text to 75 characters like described [here](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-body)? I think the CI bot will complain otherwise...

Thank you for the improvements!

yunsmall committed:
drivers: video: ov2640: Running clang-format on ov2640.c file

Using clangd to format ov2640.c file, which is not compatible with
existing .clang-format file.

Signed-off-by: Yunjie Ye <yun_small@163.com>

avolmat-st commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -160,14 +160,17 @@ Stepper
 
 Video
-===
+=====
```
Here as well, no need to modify the section title

avolmat-st reviewed (changes requested):
Thanks a lot for the added fix and addition in the migration guide. Looks good to me.

I made very small comments.

Also, regarding the 1st commit log, what about simply mentioning something like

```
drivers: video: ov2640: Add pwdn pin support on ov2640.

Add handling for the power-down (pwdn) pin of the OV2640 sensor.

Signed-off-by: yunsmall <yun_small@163.com>
```

yunsmall commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -160,14 +160,17 @@ Stepper
 ===
 
-  * :dtcompatible:`maxim,max3421e_spi` has been renamed to :dtcompatible:`maxim,max3421e-spi`.
+* :dtcompatible:`maxim,max3421e_spi` has been renamed to :dtcompatible:`maxim,max3421e-spi`.
```
Sorry, it's my fault. I saw that there were formatting issues with these things and couldn't resist fixing them.😂

yunsmall committed:
drivers: video: ov2640: Fixing the issue of the reset pin for OV2640.

The dt binding file requires the reset pin to be configured
as `GPIO_ACTIVE_LOW`, but the driver uses the
logic of `GPIO_ACTIVE_HIGH`.
Therefore, modify the logic in the driver and add a
migration guide.

Signed-off-by: Yunjie Ye <yun_small@163.com>

avolmat-st reviewed (approved):
LGTM, Thanks a lot for this work. Approved.

avolmat-st reviewed (dismissed):
LGTM.

avolmat-st commented:
@yunsmall, could you rebase this PR in order to fix conflict and be able to merge it ?

yunsmall commented:
OK.

avolmat-st reviewed (approved):
Thanks for the update. LGTM.
